Dapper.MoqTests/ScalarValue.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Dapper.MoqTests
+{
+    internal class ScalarValue : IConvertible
+    {
+        private readonly bool _isAsync;
+        private readonly MethodInfo _method;
+        private readonly object[] _parametersArray;
+        private readonly MockDatabase _mockDatabase;
+
+        public ScalarValue(bool isAsync, MethodInfo method, object[] parametersArray, MockDatabase mockDatabase)
+        {
+            _isAsync = isAsync;
+            _method = method;
+            _parametersArray = parametersArray;
+            _mockDatabase = mockDatabase;
+        }
+
+        private object Invoke()
+        {
+            return _isAsync
+                ? ((Task<object>)_method.Invoke(_mockDatabase, _parametersArray)).Result
+                : _method.Invoke(_mockDatabase, _parametersArray);
+        }
+
+        private object Invoke(Type dataType)
+        {
+            var method = _method.GetGenericMethodDefinition().MakeGenericMethod(dataType);
+            return _isAsync
+                ? ((Task<object>)method.Invoke(_mockDatabase, _parametersArray)).Result
```

Comment:
Probably need to split this into statements that permit accessing result from a 'correctly' typed instance of Task<>